plugins/outputs/mqtt/mqtt.go

```diff
@@ -91,24 +104,15 @@ func (m *MQTT) Write(metrics []telegraf.Metric) error {
 	if len(metrics) == 0 {
 		return nil
 	}
+
+	metricsmap := make(map[string][]telegraf.Metric)
 	hostname, ok := metrics[0].Tags()["host"]
 	if !ok {
 		hostname = ""
 	}
-
-	metricsmap := make(map[string][]telegraf.Metric)
-
 	for _, metric := range metrics {
-		var t []string
-		if m.TopicPrefix != "" {
-			t = append(t, m.TopicPrefix)
-		}
-		if hostname != "" {
-			t = append(t, hostname)
-		}
-
-		t = append(t, metric.Name())
-		topic := strings.Join(t, "/")
+		tt := &TemplateTopic{Hostname: hostname, metric: metric}
+		topic := tt.Parse(m)
```

Comment:
How about defining _one_ template-topic containing all the static fields like hostname, prefix, the template itself etc. This can be done in `Init()` _once_ and then just pass in the metric as a parameter to `Parse()`? Btw: I would name the function rather `Generate()` as that's what it does, _generate_ a topic...

Reply:
The hostname is read from the metric, attached as a tag.
I thought about it, that I could do it differently, but that would be a change of behaviour, I think leading to potential unintended consequences.

It may be the correct thing to do, but I wonder how safe is it for existent applications depending on this

Reply:
I don't see your point. Doing it the way I suggested, will lead to the same result (just need to also pass the `hostname` to `tt.Parse`... The difference is that you do trash a lot of memory if handling many metrics by instantiating and compile those template for _every_ single metric! Think of cases where you see 10th of thousands of those metrics...

Reply:
yes, I get it now. Thank you for that explanation 🙇